File: two-pointers/find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    def findNext(self, next: List, s:str):
        next[0]=0
        j = 0

        for i in range(1, len(s)):
            while j>0 and s[i]!=s[j]:
                j = next[j-1]
            if s[i] == s[j]:
                j+=1
            next[i] = j

    def strStr(self, haystack: str, needle: str) -> int:
        next = [0]*len(needle)
        self.findNext(next, needle)
        j = 0

        for i in range(len(haystack)):
            while j>0 and haystack[i]!=needle[j]:
                j=next[j-1]

            if haystack[i] == needle[j]:
                j+=1
            
            if j==len(needle):
                return i-len(needle)+1
        return -1

File: string/find-the-index-of-the-first-occurrence-in-a-string.cpp
class Solution {
public:
    void getNext(vector<int>& next, string& s){
        next[0]=0;
        int j=0;

        for(int i=1;i<s.size();i++){
            while (j>0 && s[i]!=s[j]){
                j=next[j-1];
            }
            if (s[i]==s[j]){
                j++;
            }
            next[i]=j;
        }
    }

    int strStr(string haystack, string needle) {
        vector<int> next(needle.size());
        getNext(next, needle);
        int j=0;

        for (int i=0; i<haystack.size(); i++){
            while (j>0 && haystack[i]!=needle[j]){
                j = next[j-1];
            }
            if (haystack[i]==needle[j]){
                j++;
            }
            if (j==needle.size()){
                return i-needle.size()+1;
            }
        }   
        return -1;
    }
};